### PR TITLE
feat: add txpool metrics

### PR DESCRIPTION
**Pull request by entropidelic**
Ref #578 

Add the following transation pool metrics:
* transaction_pool.inserted_transactions
* transaction_pool.invalid_transactions
* transaction_pool.removed_transactions

These are bundled toguether in a `TxPoolMetrics` structs.
Describers and docs were also added.

### CODE REVIEW

File: bin/reth/src/node/mod.rs [hunk 1]
```diff
@@ -95,6 +95,7 @@ impl Command {
             info!("Starting metrics endpoint at {}", listen_addr);
             prometheus_exporter::initialize(listen_addr)?;
             stages_metrics_describer::describe();
+            reth_transaction_pool::metrics::describe();
```
Comment (onbjerg) [inline bin/reth/src/node/mod.rs:98]: We should standardize how this is going to be done, one uses a module the other uses a function. Perhaps we should just have a `X::describe()` on the metrics structs if we are using those? (they are not strictly necessary btw)

File: crates/transaction-pool/src/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
+//! Transaction pool metrics.
+
+use metrics::{describe_counter, register_counter, Counter};
+
+pub(crate) struct TxPoolMetrics {
```
Comment (onbjerg) [inline crates/transaction-pool/src/metrics.rs:5]: We technically have multiple subpools, so perhaps this should be scoped by subpool? I.e. one for "ready" transactions, one where the basefee is too low, and one last I don't remember. Even more so, it is possible to have as many custom subpools as you want I think (cc @mattsse) - so really, these metrics should be scoped by subpool name

Comment (mattsse) [inline crates/transaction-pool/src/metrics.rs:5]: please add some docs, since this is public.

File: crates/transaction-pool/src/pool/txpool.rs [hunk 3]
```diff
@@ -85,6 +86,8 @@ pub struct TxPool<T: TransactionOrdering> {
     basefee_pool: ParkedPool<BasefeeOrd<T::Transaction>>,
     /// All transactions in the pool.
     all_transactions: AllTransactions<T::Transaction>,
+    /// Transaction pool metrics
+    metrics: TxPoolMetrics,
```
Comment (mattsse) [inline crates/transaction-pool/src/pool/txpool.rs:90]: can this not live on the outer level in `PoolInner`?

File: crates/transaction-pool/src/metrics.rs (hunk 2 above)
Comment (mattsse) [reply crates/transaction-pool/src/metrics.rs:5]: hmm, I think we should definitely have counters for the entire thing, and perhaps additional counters for subpools.
so starting with these is fine imo.

Comment (mattsse) [inline crates/transaction-pool/src/metrics.rs:5]: for metrics we want:

* it to be Optional
* shareabel

so I think Ideally this should look like

```rust
struct 
```TxPoolMetrics {
    inner: Arc<Option<TxPoolMetricsInner>>
}
```

where `TxPoolMetricsInner` is the type that has the counter

this way we could share it and use it in multiple pools for example

**Review by mattsse — CHANGES_REQUESTED** [review]
nice,

a few suggestions

File: crates/transaction-pool/src/metrics.rs (hunk 2 above)
Comment (onbjerg) [reply crates/transaction-pool/src/metrics.rs:5]: Afaik the metrics are all shareable and "optional" already - if metrics are disabled by the consumer, these are effectively no-ops: https://docs.rs/metrics/latest/metrics/#in-libraries

Comment (mattsse) [reply crates/transaction-pool/src/metrics.rs:5]: I see, so having Option would be an overhead if one is installed. and counters are already Option<Arc>

nvm then.

the metrics macros obfuscate a lot.

Comment (onbjerg) [reply crates/transaction-pool/src/metrics.rs:5]: They do, I do wonder if there is a more "hands on" API w/o macros

**Comment by gakonst** [comment]
Let's hold off on merging this until we merge #592, so that we dedup.

File: bin/reth/src/node/mod.rs (hunk 1 above)
Comment (entropidelic) [reply bin/reth/src/node/mod.rs]: I agree, I have changed the describers to be a method of each of metrics struct now and standardized the naming conventions. Also removed the `<crate>_metrics_describer` module

**Comment by gakonst** [comment]
@entropidelic do you perhaps want to rebase on top of #592? This will let you move your manual impls inside the type definition, and then we can easily merge this on top

**Comment by entropidelic** [comment]
> @entropidelic do you perhaps want to rebase on top of #592? This will let you move your manual impls inside the type definition, and then we can easily merge this on top

sounds good, will do. Thanks @gakonst

File: crates/transaction-pool/src/pool/txpool.rs [hunk 4]
```diff
@@ -231,14 +239,23 @@ impl<T: TransactionOrdering> TxPool<T> {
...
-                Err(PoolError::ReplacementUnderpriced(existing))
-            }
-            Err(InsertErr::ProtocolFeeCapTooLow { transaction, fee_cap }) => {
-                Err(PoolError::ProtocolFeeCapTooLow(*transaction.hash(), fee_cap))
-            }
-            Err(InsertErr::ExceededSenderTransactionsCapacity { transaction }) => {
-                Err(PoolError::SpammerExceededCapacity(transaction.sender(), *transaction.hash()))
+            Err(e) => {
+                // Update invalid transactions metric
+                self.metrics.invalid_transactions.increment(1);
```
Comment (gakonst) [inline crates/transaction-pool/src/pool/txpool.rs:244]: Do we want to be more granular here and actually have a metric on a per error basis? Maybe not, just asking the Q.

File: crates/transaction-pool/src/pool/txpool.rs (hunk 3 above)
Comment (gakonst) [reply crates/transaction-pool/src/pool/txpool.rs:90]: (why?)

File: crates/stages/src/metrics.rs [hunk 5]
```diff
@@ -10,11 +10,32 @@ pub struct HeaderMetrics {
     pub headers_timeout_errors: Counter,
     /// Number of validation errors while requesting headers
     pub headers_validation_errors: Counter,
-    /// Elapsed time of successful header requests
+    /// Number of unexpected errors while requesting headers
     pub headers_unexpected_errors: Counter,
 }
 
 impl HeaderMetrics {
```
Comment (gakonst) [inline crates/stages/src/metrics.rs:17]: Let's use the derive here

File: crates/stages/src/lib.rs [hunk 6]
```diff
@@ -40,5 +37,5 @@ pub use pipeline::*;
 pub use stage::*;
 
 // NOTE: Needed so the link in the module-level rustdoc works.
-#[allow(unused_extern_crates)]
-extern crate metrics;
+// #[allow(unused_extern_crates)]
+// extern crate metrics;
```
Comment (gakonst) [inline crates/stages/src/lib.rs:41]: remove

**Review by gakonst — CHANGES_REQUESTED** [review]
Nice work - nits

**Review by gakonst — APPROVED** [review]
Good from me!

File: crates/transaction-pool/src/pool/txpool.rs (hunk 4 above)
Comment (entropidelic) [reply crates/transaction-pool/src/pool/txpool.rs:244]: sure! I guess that tracking the exact error could be useful. I can make a separate PR for that

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**